File: 20/20.py
from timing_util import print_elapsed, timestamp_nano


def get_data(content: str) -> list[int]:
    return list(map(int, content.splitlines()))


def solve(data: list[int], encryption_key: int = None, n_mixing: int = 1) -> int:
    if encryption_key:
        data = [x * encryption_key for x in data]

    i_zero = data.index(0)
    n = len(data)

    # numbers are not unique -> generate unique key
    data = list(enumerate(data))
    copy = data.copy()
    for _ in range(n_mixing):
        for key_num in data:
            i = copy.index(key_num)
            target = i + copy.pop(i)[1]
            if -n <= target < n:
                copy.insert(target, key_num)
            else:
                copy.insert(target % (n - 1), key_num)

    i_zero = copy.index((i_zero, 0))
    return sum([copy[(i_zero + i) % n][1] for i in [1000, 2000, 3000]])


if __name__ == '__main__':
    start = timestamp_nano()

    with open('20/input.txt') as in_file:
        data = get_data(in_file.read())

    print(f'part1: {solve(data)}')
    print(f'part2: {solve(data, encryption_key=811589153, n_mixing=10)}')

    print_elapsed(start)

File: 20/20.rs
use std::fs;
use std::time::Instant;

fn main() {
    let start = Instant::now();

    let data: Vec<i64> = fs::read_to_string("20/input.txt")
        .expect("File not found")
        .lines()
        .map(|num| num.parse::<i64>().unwrap())
        .collect();

    println!("part1: {}", solve(&data, 0, 1));
    println!("part2: {}", solve(&data, 811589153, 10));
    println!("time: {:?}", start.elapsed());
}

fn solve(data: &[i64], encryption_key: i64, n_mixing: usize) -> i64 {
    let mut i_zero = data.iter().position(|&x| x == 0).unwrap();
    let numbers: Vec<(usize, i64)> = if encryption_key != 0 {
        data.iter()
            .map(|&x| x * encryption_key)
            .enumerate()
            .collect()
    } else {
        data.iter().copied().enumerate().collect()
    };

    let mut ring: Vec<(usize, i64)> = numbers.clone();
    let n: i64 = numbers.len() as i64;
    for _ in 0..n_mixing {
        for key_num in numbers.iter() {
            let mut i: i64 = ring.iter().position(|entry| entry == key_num).unwrap() as i64;
            i += ring.remove(i as usize).1;

            if !(-n <= i && i < n) {
                i %= n - 1;
            }
            if -n <= i && i < 0 {
                i += n - 1;
            }

            ring.insert(i as usize, *key_num);
        }
    }

    i_zero = ring.iter().position(|&elem| elem == (i_zero, 0)).unwrap();
    (1000..=3000)
        .step_by(1000)
        .map(|i| ring[(i_zero + i) % ring.len()].1)
        .sum()
}
